protocols/relay/src/v2/protocol/outbound_hop.rs

```diff
@@ -0,0 +1,221 @@
+        let msg = match self {
+            Upgrade::Reserve => HopMessage {
+                r#type: hop_message::Type::Reserve.into(),
+                peer: None,
+                reservation: None,
+                limit: None,
+                status: None,
+            },
+            Upgrade::Connect { dst_peer_id } => HopMessage {
+                r#type: hop_message::Type::Connect.into(),
+                peer: Some(Peer {
+                    id: dst_peer_id.to_bytes(),
+                    addrs: vec![],
+                }),
+                reservation: None,
+                limit: None,
+                status: None,
+            },
+        };
+
+        let mut encoded_msg = Vec::new();
+        msg.encode(&mut encoded_msg)
+            .expect("Vec to have sufficient capacity.");
+
+        let mut codec = UviBytes::default();
+        codec.set_max_len(MAX_MESSAGE_SIZE);
+        let mut substream = Framed::new(substream, codec);
+
+        async move {
+            substream.send(std::io::Cursor::new(encoded_msg)).await?;
+            let msg: bytes::BytesMut = substream
+                .next()
+                .await
+                .ok_or_else(|| std::io::Error::new(std::io::ErrorKind::UnexpectedEof, ""))??;
+
+            let HopMessage {
+                r#type,
+                peer: _,
+                reservation,
+                limit: _,
+                status,
+            } = HopMessage::decode(Cursor::new(msg))?;
+
+            let r#type = hop_message::Type::from_i32(r#type).ok_or(UpgradeError::ParseTypeField)?;
+            match r#type {
+                hop_message::Type::Connect => return Err(UpgradeError::UnexpectedTypeConnect),
+                hop_message::Type::Reserve => return Err(UpgradeError::UnexpectedTypeReserve),
+                hop_message::Type::Status => {}
+            }
+
+            let status = Status::from_i32(status.ok_or(UpgradeError::MissingStatusField)?)
+                .ok_or(UpgradeError::ParseStatusField)?;
+            match status {
+                Status::Ok => {}
+                s => return Err(UpgradeError::UnexpectedStatus(s)),
+            }
+
+            let output = match self {
+                Upgrade::Reserve => {
+                    let reservation = reservation.ok_or(UpgradeError::MissingReservationField)?;
+
+                    let addrs = if reservation.addrs.is_empty() {
+                        return Err(UpgradeError::NoAddressesInReservation);
+                    } else {
+                        reservation
+                            .addrs
+                            .into_iter()
+                            .map(TryFrom::try_from)
+                            .collect::<Result<Vec<Multiaddr>, _>>()
+                            .map_err(|_| UpgradeError::InvalidReservationAddrs)?
+                    };
+
+                    let renewal_timeout = if let Some(timestamp) = reservation.expire {
```

Comment:
Is the expiration timestamp really optional? In our implementation in `inbound_hop::ReservationReq::accept` it is always set.
The specs aren't clear on this either:

> the expire field contains the expiration time as a UTC UNIX time. The reservation becomes invalid after this time and it's the responsibility of the client to refresh.

Reply:
Thanks for raising this. That was a bit of a rabbit hole. Moved the discussion to https://github.com/libp2p/specs/pull/384.

Reply:
https://github.com/libp2p/specs/pull/384/ is merged and d5ee477b92a9342a6f2cb98462f5a58526348bcb applies the corresponding changes to this pull request.